lens/database.ml

```diff
@@ -302,5 +336,78 @@ module Insert = struct
       (fmt_vals |> Format.pp_comma_list)
       v.values fmt_returning ()
 
+  let split v =
+    let { table; columns; values; returning } = v in
+    let f v =
+      let values = [ v ] in
+      { table; columns; values; returning }
+    in
+    List.map ~f values
+
+  let no_returning v = { v with returning = [] }
+
+  let exec_insert_returning_hack ~db data =
+    let last_id_fun =
+      let driver = db.driver_name () in
+      match driver with
+      | "mysql" -> "last_insert_id()"
+      | "sqlite3" -> "last_insert_rowid()"
+      | _ ->
+          let fn = "exec_insert_returning_hack" in
+          E.Unsupported_for_driver { driver; fn } |> E.raise
+    in
+    let getid = Format.asprintf "SELECT %s as id" last_id_fun in
+    let exec_insert v =
+      let ins = Format.asprintf "%a" (fmt ~db) v in
+      Debug.print ins;
+      Statistics.time_query (fun () -> db.execute ins);
+      Debug.print getid;
+      Statistics.time_query (fun () ->
+          db.execute_select getid ~field_types:[ ("id", Phrase_type.Serial) ])
+    in
+    no_returning data |> split |> List.map ~f:exec_insert |> List.concat
+
+  let exec_insert_returning ~db ~field_types data =
+    match db.driver_name () with
+    | "sqlite3"
+     |"mysql" ->
```

Comment:
indentation/spacing

Reply:
This is part of the current ocamlformat settings, where the alternative syntax is used to highlight that fact that the second case is a fallthrough. I think this setting should be kept, or at the very least not changed in this PR. The docs for this formatting behaviour:

```
       --indicate-nested-or-patterns={unsafe-no|space}
           Control whether or not to indicate nested or-pattern using
           indentation. unsafe-no does not indicate nested or-patterns.
           Warning: this can produce confusing code where a short body of a
           match case is visually hidden by surrounding long patterns, leading
           to misassociation between patterns and body expressions. space
           starts lines of nested or-patterns with " |" rather than "| ". The
           default value is unsafe-no.
```

Reply:
Oh, OK.